src/nfa.rs

```diff
@@ -322,6 +323,42 @@ impl<S: StateID> State<S> {
     }
 }
 
+#[derive(Clone, Debug)]
+struct Dense<S>(Vec<S>);
+
+impl<S> Dense<S>
+    where S: StateID,
+{
+    fn new() -> Self {
+        Dense(vec![fail_id(); 256])
+    }
+
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+impl<S> Index<u8> for Dense<S> {
+    type Output = S;
+
+    #[inline(always)]
+    fn index(&self, i: u8) -> &S {
+        // SAFETY: This is safe because all dense transitions have
+        // exactly 256 elements, so all u8 values are valid indices.
+        unsafe { self.0.get_unchecked(i as usize) }
+    }
+}
+
+impl<S> IndexMut<u8> for Dense<S> {
+    #[inline(always)]
+    fn index_mut(&mut self, i: u8) -> &mut S {
+        // SAFETY: This is safe because all dense transitions have
+        // exactly 256 elements, so all u8 values are valid indices.
+        unsafe { self.0.get_unchecked_mut(i as usize) }
+    }
+}
```

Comment:
Ah this is great! (I am also kind of curious whether using `unsafe` here is necessary at all. It feels like a bounds check that LLVM should be able to elide.)

Reply:
Yeah, it is unlikely to have an effect after LLVM is done with the code. It is one less thing for it to optimize I guess 🤷‍♂️

Reply:
Aye, next time I think of it I'll check out the before/after benchmarks with the unsafe removed here. If I can't see a difference, then I'll drop it.